File: app/schema/hazard.py
"""隐患治理闭环 Schema：创建/更新/详情/状态流转/分页/统计。

时间字段统一以「北京时间墙钟」字符串对外（YYYY-MM-DDTHH:mm:ss），与前端
el-date-picker 的 value-format 对齐；入库存 timestamptz（约定按 Asia/Shanghai 解释）。
"""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_serializer

from app.core.clock import LOCAL_TZ
from app.core.constants import (
    HAZARD_CATEGORIES,
    HAZARD_LEVELS,
    HAZARD_SOURCES,
    HAZARD_STATUSES,
)


class HazardCreate(BaseModel):
    project_id: int | None = None
    title: str = Field(..., description="隐患标题")
    level: str = Field("一般", description="隐患等级(重大/较大/一般/低)")
    category: str | None = Field(None, description="隐患类别")
    description: str | None = None
    location_desc: str | None = None
    lng: float | None = None
    lat: float | None = None
    discovered_by_name: str | None = None
    discovered_at: datetime | None = None
    source: str = "人工"
    assignee_id: int | None = None
    due_at: datetime | None = None


class HazardUpdate(BaseModel):
    project_id: int | None = None
    title: str | None = None
    level: str | None = None
    category: str | None = None
    description: str | None = None
    location_desc: str | None = None
    lng: float | None = None
    lat: float | None = None
    discovered_by_name: str | None = None
    discovered_at: datetime | None = None
    source: str | None = None
    assignee_id: int | None = None
    due_at: datetime | None = None


class HazardTransition(BaseModel):
    action: str = Field(
        ...,
        description="流转动作(start_rectify/submit_rectify/verify_pass/verify_reject/reject/reopen)",
    )
    note: str | None = Field(None, description="流转说明(整改说明/复核意见/驳回原因)")


class HazardOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    project_id: int | None = None
    project_name: str | None = None
    title: str
    level: str = "一般"
    category: str | None = None
    description: str | None = None
    location_desc: str | None = None
    lng: float | None = None
    lat: float | None = None
    discovered_by_name: str | None = None
    discovered_at: datetime | None = None
    source: str = "人工"
    source_alarm_id: int | None = None
    status: str = "待整改"
    assignee_id: int | None = None
    assignee_name: str | None = None
    due_at: datetime | None = None
    rectify_note: str | None = None
    rectify_at: datetime | None = None
    verify_by_name: str | None = None
    verify_at: datetime | None = None
    verify_note: str | None = None
    closed_at: datetime | None = None
    created_by: int | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None
    is_overdue: bool = False

    @field_serializer(
        "discovered_at",
        "due_at",
        "rectify_at",
        "verify_at",
        "closed_at",
        "created_at",
        "updated_at",
    )
    def _serialize_dt(self, v: datetime | None) -> str | None:
        """带时区时间序列化为「北京时间墙钟」字符串（YYYY-MM-DDTHH:mm:ss）。"""
        if v is None:
            return None
        return v.astimezone(LOCAL_TZ).replace(tzinfo=None).isoformat()


class HazardPage(BaseModel):
    total: int
    items: list[HazardOut]
    page: int = 1
    size: int = 20


class HazardStats(BaseModel):
    total: int
    by_status: dict[str, int]
    by_level: dict[str, int]
    overdue: int


# 供前端下拉复用的枚举
HAZARD_LEVEL_OPTIONS = list(HAZARD_LEVELS)
HAZARD_CATEGORY_OPTIONS = list(HAZARD_CATEGORIES)
HAZARD_SOURCE_OPTIONS = list(HAZARD_SOURCES)
HAZARD_STATUS_OPTIONS = list(HAZARD_STATUSES)

File: web/src/api/hazard.ts
// 隐患治理闭环 API 封装
import request, { http } from "@/utils/request";

export type HazardLevel = "重大" | "较大" | "一般" | "低";
export type HazardStatus =
  | "待整改"
  | "整改中"
  | "待复核"
  | "已销号"
  | "已驳回";

export interface Hazard {
  id: number;
  project_id: number | null;
  project_name: string | null;
  title: string;
  level: HazardLevel;
  category: string | null;
  description: string | null;
  location_desc: string | null;
  lng: number | null;
  lat: number | null;
  discovered_by_name: string | null;
  discovered_at: string | null;
  source: string;
  status: HazardStatus;
  assignee_id: number | null;
  assignee_name: string | null;
  due_at: string | null;
  rectify_note: string | null;
  rectify_at: string | null;
  verify_by_name: string | null;
  verify_at: string | null;
  verify_note: string | null;
  closed_at: string | null;
  created_by: number | null;
  created_at: string | null;
  updated_at: string | null;
  is_overdue: boolean;
}

export interface HazardCreate {
  project_id?: number | null;
  title: string;
  level?: HazardLevel;
  category?: string | null;
  description?: string | null;
  location_desc?: string | null;
  lng?: number | null;
  lat?: number | null;
  discovered_by_name?: string | null;
  discovered_at?: string | null;
  source?: string;
  assignee_id?: number | null;
  due_at?: string | null;
}

export type HazardUpdate = Partial<HazardCreate>;

export interface HazardTransition {
  action: string;
  note?: string | null;
}

export interface HazardStats {
  total: number;
  by_status: Record<string, number>;
  by_level: Record<string, number>;
  overdue: number;
}

export interface HazardOptions {
  levels: string[];
  categories: string[];
  sources: string[];
  statuses: string[];
}

export interface HazardPage {
  total: number;
  items: Hazard[];
  page: number;
  size: number;
}

// 分页列表（带筛选 + 数据隔离）
export function fetchHazards(params: {
  project_id?: number;
  level?: string;
  status?: string;
  keyword?: string;
  overdue?: boolean;
  page?: number;
  size?: number;
}): Promise<HazardPage> {
  return http<HazardPage>({ url: "/v1/hazards", method: "GET", params });
}

export function createHazard(req: HazardCreate): Promise<Hazard> {
  return http<Hazard>({ url: "/v1/hazards", method: "POST", data: req });
}

export function updateHazard(id: number, req: HazardUpdate): Promise<Hazard> {
  return http<Hazard>({ url: `/v1/hazards/${id}`, method: "PUT", data: req });
}

export function deleteHazard(id: number): Promise<{ id: number }> {
  return http<{ id: number }>({ url: `/v1/hazards/${id}`, method: "DELETE" });
}

export function transitionHazard(id: number, req: HazardTransition): Promise<Hazard> {
  return http<Hazard>({
    url: `/v1/hazards/${id}/transition`,
    method: "POST",
    data: req,
  });
}

export function fetchHazardStats(): Promise<HazardStats> {
  return http<HazardStats>({ url: "/v1/hazards/stats", method: "GET" });
}

export function fetchHazardOptions(): Promise<HazardOptions> {
  return http<HazardOptions>({ url: "/v1/hazards/options", method: "GET" });
}

// 导出隐患报表（excel|pdf）——返回二进制 Blob
export interface HazardExportParams {
  project_id?: number;
  level?: string;
  status?: string;
  keyword?: string;
  overdue?: boolean;
}
export async function exportHazardReport(
  fmt: "excel" | "pdf",
  params: HazardExportParams,
): Promise<Blob> {
  const resp = await request.get("/v1/hazards/export", {
    params: { ...params, fmt },
    responseType: "blob",
  });
  return resp.data as Blob;
}
